lib/fine-types/src/Language/FineTypes/Typ/Gen.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Language.FineTypes.Typ.Gen where
+
+import Prelude
+
+import Data.Char (toUpper)
+import Data.List (nub)
+import Data.Traversable (forM)
+import Language.FineTypes.Typ
+    ( ConstructorName
+    , FieldName
+    , OpOne (..)
+    , OpTwo (..)
+    , Typ (Abstract, One, ProductN, SumN, Two, Var, Zero)
+    , TypConst (..)
+    , TypName
+    )
+import Test.QuickCheck
+    ( Gen
+    , elements
+    , listOf1
+    , oneof
+    , scale
+    , vectorOf
+    )
+
+-- | If the generated 'Typ' should be concrete or not. 'Concrete' will not contain
+-- 'Abstract' or 'Var' leaves
+data Concrete = Complete | Concrete
+
+patchNoData :: Concrete -> [Gen Typ] -> [Gen Typ]
+patchNoData Concrete = id
+patchNoData Complete =
+    (<>)
+        [ pure Abstract
+        , Var <$> genVarName
+        ]
+
+-- | Minimum depth of the generated 'Typ'. Shorter than depth branches are still
+-- possible if the actual  'Typ' is Zero or Abstract or Var
+newtype DepthGen = DepthGen Int
+    deriving (Eq, Ord, Show, Num)
+
+shaping :: DepthGen -> Gen Bool
+shaping (DepthGen n)
+    | n > 0 = pure True
+    | otherwise = elements $ True : replicate (negate n) False
```

Comment:
??? What does this function do? 🤔 We have `replicate (negate n) = []` for all positive `n`.

Reply:
Ah, `n <= 0` in this branch, sorry. 🙈 

Hm. Perhaps write it more explicitly like this? 🤔 

```hs
willBranch :: DepthGen -> Gen Bool
willBranch (DepthGen n)
    | n > 0 = pure True
    | otherwise = frequency [(1,True), (negative n, False)]
```

`willBranch` might be a better name for it.